### ISSUE TITLE
threshold_li fails for "zeros" input images due to undefined `threshold`

### ISSUE THREAD

**Issue by jmetz** [scikit-image/scikit-image#2337]
## Description

When running `skimage.filters.threshold_li`, if the input image is all 0, the 
`tolerance` is set to zero causing the while loop never to execute and a somewhat cryptic error is
encountered because `threshold` never gets set. 

A typical traceback is 

```
Traceback (most recent call last):
  File "test.py", line 3, in <module>
    threshold_li(np.zeros((100,100)))
  File "./skimage/filters/thresholding.py", line 509, in threshold_li
    return threshold + immin
UnboundLocalError: local variable 'threshold' referenced before assignment
```

I recognise that this is a special case, but I just wanted to raise this issue as a way of
asking 

> do we want this somewhat cryptic error, or should we catch this 
> using a check on the input image"?  
## Way to reproduce

All that's needed to reproduce this (Python 3.4.1, skimage 0.13dev) is: 

```
import numpy as np
from skimage.filters import threshold_li
threshold_li(np.zeros((100,100)))
```

**Comment by soupault** [comment]
Closed via #2338.

**Comment by jni** [comment]
@stefanv huh? Why did you reopen?

**Comment by soupault** [comment]
Probably, is a missclick. I _actually_ do think that this is closed in #2338.
Please, feel free to reopen with an explanation.

**Comment by jni** [comment]
A zombie issue coming back to haunt us!

```python
import numpy as np
from skimage import filters
filters.threshold_li(np.zeros((5, 5)) / 0)
```

```pytb
/home/jni/miniconda3/envs/cf/bin/ipython:1: RuntimeWarning: invalid value encountered in true_divide
  #!/home/jni/miniconda3/envs/cf/bin/python
/home/jni/miniconda3/envs/cf/lib/python3.6/site-packages/numpy/core/_methods.py:29: RuntimeWarning: invalid value encountered in reduce
  return umr_minimum(a, axis, None, out, keepdims)
/home/jni/miniconda3/envs/cf/lib/python3.6/site-packages/numpy/core/_methods.py:26: RuntimeWarning: invalid value encountered in reduce
  return umr_maximum(a, axis, None, out, keepdims)
---------------------------------------------------------------------------
UnboundLocalError                         Traceback (most recent call last)
<ipython-input-11-e57608735ae9> in <module>()
----> 1 filters.threshold_li(np.zeros((5, 5)) / 0)

~/projects/scikit-image/skimage/filters/thresholding.py in threshold_li(image)
    532             new_thresh = temp + tolerance
    533 
--> 534     return threshold + immin
    535 
    536 

UnboundLocalError: local variable 'threshold' referenced before assignment
```


Discovered while working on a failing CellProfiler pipeline. Fun fact: the source of the nan image was also us:

```python
>>> import numpy as np
>>> from skimage import exposure
>>> exposure.rescale_intensity(np.zeros((5, 5))
/home/jni/projects/scikit-image/skimage/exposure/exposure.py:296: RuntimeWarning: invalid value encountered in true_divide
  image = (image - imin) / float(imax - imin)
array([[nan, nan, nan, nan, nan],
       [nan, nan, nan, nan, nan],
       [nan, nan, nan, nan, nan],
       [nan, nan, nan, nan, nan],
       [nan, nan, nan, nan, nan]])

``` 

@bethac07 suggests that we could return the same image back if the range is 0 in rescale intensity. Thoughts?

**Comment by jmetz** [comment]
@jni - this is because 

```python
    np.nan == np.nan
```

returns `False` (which is essentially the check I added in #2338 ). 

I don't think checking the range will work here as expected, or at least e.g. `np.ptp` returns `nan` on a `nan` input array. 

Also, even if we were to use `nan`-aware functions, such as `nanmin` or `nanmax`, most threshold functions would likely still choke with **any** `nan` elements.  

I would suggest checking for any `nan` (e.g. `np.any(np.isnan(image))`) and handling that explicitly.

**Comment by jni** [comment]
@jmetz yeah I traced it back to that. Honestly in addition to fixing this, I want to propose that we stop raising ValueErrors on flat images, and instead return the single value of the image. (In this case, it would be `np.nan`. But in cases with single nans, they would be ignored.

As a side note, in addition to the above check, `UnboundLocalError` signals an antipattern. In this case, it should be impossible to reach the return statement without declaring threshold. It should be set to *some* value above the while loop.

**Comment by jmetz** [comment]
@jni it's not clear to me what the most sensible treatment of flat images would be, though returning the single value sounds like a reasonable approach. 

If `nan`s interspersed with non-`nan`s should be ignored we could 

a) use `np.ma.masked_invalid` to return a masked version of the image with all invalid values masked
b) go more basic and perform something like `image = image[np.isfinite(image)]` 

I totally agree regarding the `UnboundLocalError` - in fact you've uncovered a bug in my port of the algorithm (https://github.com/fiji/Auto_Threshold/blob/Auto_Threshold-1.16.5/src/main/java/fiji/threshold/Auto_Threshold.java) - the ImageJ version used a `do-while` which I've replaced by a `while`. 

In summary, I'd suggest that perhaps the points raised constitute **2 new issues** instead of re-opening and adding to this issue. Basically they should be something like

a) Threshold functions not handling `nan`s in images. 
b) `threshold_li` `while` loop fails to execute causing `UnboundLocalError`. 

We could further split a) into the case of all `nan`s vs some `nan`s if desired - I'm not sure what level of granularity is preferred here?

**Comment by stefanv** [comment]
On Mon, 10 Sep 2018 14:42:04 +0000, Jeremy Metz wrote:
> @jni it's not clear to me what the most sensible treatment of flat
> images would be, though returning the single value sounds like a
> reasonable approach.

I cannot think of any other return value that would make sense.  And
throwing an error feels unnecessary.  The user can also interpret that
threshold in a way that makes sense:

```
image > T   # all False
image <= T  # all True
```

> a) Threshold functions not handling `nan`s in images. 
> b) `threshold_li` `while` loop fails to execute causing
> `UnboundLocalError`.

+1 for creating these two issues and addressing them separately,
possibly (b) first.
